reana_client/validation/environments.py

```diff
@@ -0,0 +1,252 @@
+    _get_image_uid_gids,
+    _get_full_image_name,
+)
+
+from reana_client.printer import display_message
+
+
+def validate_environment(reana_yaml, pull=False):
+    """Validate environments in REANA specification file according to workflow type.
+
+    :param reana_yaml: Dictionary which represents REANA specifications file.
+    :param pull: If true, attempt to pull remote environment image to perform GID/UID validation.
+    """
+
+    def build_validator(workflow):
+        workflow_type = workflow["type"]
+        if workflow_type == "serial":
+            workflow_steps = workflow["specification"]["steps"]
+            return SerialEnvironmentValidator(workflow_steps=workflow_steps, pull=pull)
+        if workflow_type == "yadage":
+            workflow_steps = workflow["specification"]["stages"]
+            return YadageEnvironmentValidator(workflow_steps=workflow_steps, pull=pull)
+        if workflow_type == "cwl":
+            workflow_file = workflow.get("file")
+            return CWLEnvironmentValidator(workflow_file=workflow_file, pull=pull)
+
+    workflow = reana_yaml["workflow"]
+    validator = build_validator(workflow)
+    validator.validate()
+    validator.display_messages()
+
+
+class EnvironmentValidatorBase:
+    """REANA workflow environments validation base class."""
+
+    def __init__(self, workflow_steps=None, workflow_file=None, pull=False):
+        """Validate environments in REANA workflow.
+
+        :param workflow_steps: List of dictionaries which represents different steps involved in workflow.
+        :param workflow_file: Path to workflow specification.
+        :param pull: If true, attempt to pull remote environment image to perform GID/UID validation.
+        """
+        self.workflow_steps = workflow_steps
+        self.workflow_file = workflow_file
+        self.pull = pull
+        self.validated_images = set()
+        self.messages = []
+
+    def validate(self):
+        """Validate REANA workflow environments."""
+        try:
+            self.validate_environment()
+        except EnvironmentValidationError as e:
+            self.messages.append({"type": "error", "message": str(e)})
+            self.display_messages()
+            sys.exit(1)
+
+    def validate_environment(self):
+        """Validate environments in REANA workflow."""
+        raise NotImplementedError
+
+    def display_messages(self):
+        """Display messages in console."""
+        for msg in self.messages:
+            display_message(
+                msg["message"], msg_type=msg["type"], indented=True,
+            )
+
+    def _validate_environment_image(self, image, kubernetes_uid=None):
+        """Validate image environment.
+
+        :param image: Full image name with tag if specified.
+        :param kubernetes_uid: Kubernetes UID defined in workflow spec.
+        """
+
+        if image not in self.validated_images:
+            image_name, image_tag, message = _validate_image_tag(image)
```

Comment:
Initially I moved all these functions inside the class, but it got really big and messy. Then I tried to move some of them to `utils`, I guess a main criteria that if it can be detached from the class quite easily, then move to `utils`, but if the function was using too many class attributes, then leave it as a class method. But I agree that it's a bit random

Reply:
In my opinion, if the functions are only used within the classes it makes sense to have them as class methods. If there are used somewhere else then it makes sense to have them in utils.

Reply:
Thanks for the feedback, moved these functions to the base class